File: ruby/rails/app/controllers/store_controller.rb
class StoreController < ApplicationController
  def get_inventory
    result = repo.get_inventory
    render json: result
  rescue => e
    render json: { message: e.message }, status: :internal_server_error
  end

  def place_order
    order = parsed_body
    result = repo.place_order(order)
    render json: result, status: :created
  rescue => e
    render json: { message: e.message }, status: :bad_request
  end

  def get_order_by_id
    order = repo.find_order_by_id(params[:orderId])
    render json: order
  rescue NotFoundError => e
    render json: { message: e.message }, status: :not_found
  end

  def delete_order
    repo.delete_order(params[:orderId])
    render json: {}
  rescue => e
    render json: { message: e.message }, status: :bad_request
  end
end

File: petshop-stacks/elixir/phoenix/lib/petstore_web/controllers/store_controller.ex
defmodule PetstoreWeb.StoreController do
  use PetstoreWeb, :controller
  import PetstoreWeb.RepoHelper

  # GET /store/inventory
  def get_inventory(conn, _params) do
    case repo().get_inventory() do
      {:ok, inventory} -> json(conn, inventory)
      {:error, reason} -> send_error(conn, 500, reason)
    end
  end

  # POST /store/order
  def place_order(conn, params) do
    case repo().place_order(params) do
      {:ok, order} -> json(conn, order)
      {:error, reason} -> send_error(conn, 400, reason)
    end
  end

  # GET /store/order/:orderId
  def get_order_by_id(conn, %{"orderId" => order_id}) do
    case Integer.parse(order_id) do
      {id, ""} ->
        case repo().get_order_by_id(id) do
          {:ok, order} -> json(conn, order)
          {:error, :not_found} -> send_error(conn, 404, "Order not found")
          {:error, reason} -> send_error(conn, 400, reason)
        end

      _ ->
        send_error(conn, 400, "Invalid order ID")
    end
  end

  # DELETE /store/order/:orderId
  def delete_order(conn, %{"orderId" => order_id}) do
    case Integer.parse(order_id) do
      {id, ""} ->
        case repo().delete_order(id) do
          :ok -> send_resp(conn, 200, "")
          {:error, :not_found} -> send_error(conn, 404, "Order not found")
          {:error, reason} -> send_error(conn, 400, reason)
        end

      _ ->
        send_error(conn, 400, "Invalid order ID")
    end
  end

  defp send_error(conn, status, message) when is_binary(message) do
    conn
    |> put_status(status)
    |> json(%{"message" => message})
  end

  defp send_error(conn, status, reason) do
    send_error(conn, status, inspect(reason))
  end
end
